File: sdk/dart/dovahlink_client/lib/src/protocol/hello_payload.dart
import 'package:json_annotation/json_annotation.dart';

import 'package:dovahlink_client_sdk/src/protocol/hello_auth_payload.dart';
import 'package:dovahlink_client_sdk/src/protocol/json_map.dart';
import 'package:dovahlink_client_sdk/src/shared/enums.dart';

part 'hello_payload.g.dart';

/// Outgoing `hello` payload (`protocol/schema/README.md`'s `hello`). Encode-only: the client
/// never decodes its own `hello`.
@JsonSerializable(createFactory: false, explicitToJson: true)
class HelloPayload {
  /// Creates a hello payload. [authToken] is required for [AuthMethod.oneTimeLocalToken] and
  /// [AuthMethod.trustedDeviceCredential], and must be absent (not merely null) for
  /// [AuthMethod.unpaired] -- a contextual invariant this constructor validates directly, since
  /// generated serialization only knows how to encode a field's declared shape, not the
  /// relationship between two of them.
  HelloPayload({
    required this.clientId,
    required AuthMethod authMethod,
    String? authToken,
  }) : auth = HelloAuthPayload(method: authMethod, token: authToken),
       assert(
         (authMethod == AuthMethod.unpaired) == (authToken == null),
         'authToken must be absent for unpaired, and present for every other auth method.',
       );

  /// The fixed endpoint identity every `hello` this client sends carries.
  final String endpoint = 'client';

  /// The logical client/installation identity, independent of any connection.
  final String clientId;

  /// The auth method and credential presented in `auth`.
  final HelloAuthPayload auth;

  /// Encodes this payload as a JSON object.
  JsonMap toJson() => _$HelloPayloadToJson(this);
}

File: integration/DovahLinkValidationClient/HelloPayload.cs
using System.Text.Json.Nodes;

namespace DovahLinkValidationClient;

/// <summary>
/// Outgoing <c>hello</c> payload (<c>protocol/schema/README.md</c>'s <c>hello</c>). Encode-only:
/// this client never decodes its own <c>hello</c>.
/// </summary>
/// <param name="ClientId">Identifies the logical client/installation, independent of any
/// connection.</param>
/// <param name="Auth">The nested authentication object.</param>
public sealed record HelloPayload(string ClientId, HelloAuthPayload Auth)
{
    /// <summary>The sender's role. Always <c>"client"</c>: only the connecting client sends
    /// <c>hello</c>.</summary>
    public string Endpoint => "client";

    /// <summary>
    /// Encodes this payload as a JSON object.
    /// </summary>
    /// <returns>The encoded <c>hello</c> payload.</returns>
    public JsonObject Encode() => new()
    {
        ["endpoint"] = Endpoint,
        ["clientId"] = ClientId,
        ["auth"] = Auth.Encode(),
    };
}
